signer/src/network/mod.rs

```diff
@@ -26,6 +37,80 @@ pub trait MessageTransfer {
+    signal_tx: Sender<SignerSignal>,
+    signal_rx: Receiver<SignerSignal>,
+}
+
+impl SignallingNetwork {
+    /// Create a new broadcast channel network instance. This requires an active
+    /// [`Context`] and will retrieve its own signalling sender and receiver.
+    pub fn new(ctx: &impl Context) -> Self {
+        Self {
+            signal_tx: ctx.get_signal_sender(),
+            signal_rx: ctx.get_signal_receiver(),
+        }
+    }
+}
+
+impl MessageTransfer for SignallingNetwork {
+    type Error = crate::error::Error;
+
+    /// This will broadcast the message to the application signalling channel
+    /// using a [`SignerCommand::P2PPublish`] command. This implementation does
+    /// not actually send the message to the P2P network, but rather signals
+    /// to the active network implementation to do so.
+    ///
+    /// Note that this is not a blocking operation, and this method will return
+    /// as soon as the message has been sent to the signalling channel.
+    ///
+    /// If you need to wait for a receipt (success/fail), you can use your own
+    /// [`Receiver<SignerSignal>`] to listen for the
+    /// [`SignerEvent::P2PPublishFailure`] and [`SignerEvent::P2PPublishSuccess`]
+    /// events, which will provide you with the [`MsgId`] to match against your
+    /// in-flight requests.
+    async fn broadcast(&mut self, msg: Msg) -> Result<(), Self::Error> {
+        self.signal_tx
+            .send(SignerSignal::Command(SignerCommand::P2PPublish(msg)))
+            .map_err(|error| error.into())
+            .map(|_| ())
+    }
+
+    /// This will listen for incoming messages on the application signalling
+    /// channel, and return the message once it has been received.
+    ///
+    /// This is a blocking operation, and will wait until a message has been
+    /// received before returning.
+    ///
+    /// ### Important Note
+    /// To avoid ending up in a slow-receiver situation, you should queue
+    /// messages in a local buffer (i.e. [`std::collections::VecDeque`]) and
+    /// process them in your own time. Otherwise, if there are a large number
+    /// of messages being sent, you risk lagging and eventually having the tail
+    /// of the receiver being dropped, thus missing messages.
+    ///
+    /// In other words, you should be calling this method as rapidly as possible.
+    async fn receive(&mut self) -> Result<Msg, Self::Error> {
+        loop {
+            match self.signal_rx.recv().await {
+                // We are only interested in received P2P messages
+                Ok(SignerSignal::Event(SignerEvent::P2PMessageReceived(msg))) => {
+                    return Ok(msg);
+                }
+                // And if we get an error when attempting to read from the
+                // channel.
+                Err(error) => {
+                    return Err(error.into());
+                }
+                // Anything else, we ignore.
+                _ => continue,
```

Comment:
That was because we're only interested in `P2PMessageReceived` events, i.e. `receive()` doesn't need to know about `P2PPublish` commands coming through.

But you made me realize that this would never shutdown on its own so I've refactored it to a `select!` instead (which is also implicitly ignoring other message types we're not interested in).

Reply:
Actually just removed that err match as it will never be hit